devex/vscode-extension/assets/edgeSolutionTemplateFolder/modules/[[project-name]]/host/main.c

```diff
@@ -0,0 +1,235 @@
+    IoTHubMessage_Destroy(messageInstance->messageHandle);
+    free(messageInstance);
+}
+
+// Allocates a context for callback and clones the message
+// NOTE: The message MUST be cloned at this stage.  InputQueue1Callback's caller always frees the message
+// so we need to pass down a new copy.
+static MESSAGE_INSTANCE* CreateMessageInstance(char* messageBody)
+{
+    MESSAGE_INSTANCE* messageInstance = (MESSAGE_INSTANCE*)malloc(sizeof(MESSAGE_INSTANCE));
+    if (NULL == messageInstance)
+    {
+        printf("Failed allocating 'MESSAGE_INSTANCE' for pipelined message\r\n");
+    }
+    else
+    {
+        memset(messageInstance, 0, sizeof(*messageInstance));
+
+        if ((messageInstance->messageHandle = IoTHubMessage_CreateFromString(messageBody)) == NULL)
+        {
+            free(messageInstance);
+            messageInstance = NULL;
+        }
+        else
+        {
+            IoTHubMessage_SetContentTypeSystemProperty(messageInstance->messageHandle, "application%2fjson");
+            IoTHubMessage_SetContentEncodingSystemProperty(messageInstance->messageHandle, "utf-8");
+
+            messageInstance->messageTrackingId = messagesReceivedByInput1Queue;
+        }
+    }
+
+    return messageInstance;
+}
+
+static IOTHUBMESSAGE_DISPOSITION_RESULT SendEnclaveResponse(IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle, char* messageBodyStr)
+{
+    IOTHUBMESSAGE_DISPOSITION_RESULT result;
+    IOTHUB_CLIENT_RESULT clientResult;
+
+    char* enclaveMessage = (char*)malloc(512 * sizeof(char));
+    int enclaveResult = call_enclave(messageBodyStr, enclaveMessage, 512);
+    if (enclaveResult != 0)
+    {
+        result = IOTHUBMESSAGE_ABANDONED;
+    }
+    else
+    {
+        // This message should be sent to next stop in the pipeline, namely "output1".  What happens at "outpu1" is determined
+        // by the configuration of the Edge routing table setup.
+        MESSAGE_INSTANCE *messageInstance = CreateMessageInstance(enclaveMessage);
+        if (NULL == messageInstance)
+        {
+            result = IOTHUBMESSAGE_ABANDONED;
+        }
+        else
+        {
+            printf("Sending message (%zu) to the next stage in pipeline\n", messagesReceivedByInput1Queue);
+
+            clientResult = IoTHubModuleClient_LL_SendEventToOutputAsync(iotHubModuleClientHandle, messageInstance->messageHandle, "output1", SendConfirmationCallback, (void *)messageInstance);
+            if (clientResult != IOTHUB_CLIENT_OK)
+            {
+                IoTHubMessage_Destroy(messageInstance->messageHandle);
+                free(messageInstance);
+                printf("IoTHubModuleClient_LL_SendEventToOutputAsync failed on sending msg#=%zu, err=%d\n", messagesReceivedByInput1Queue, clientResult);
+                result = IOTHUBMESSAGE_ABANDONED;
+            }
+            else
+            {
+                result = IOTHUBMESSAGE_ACCEPTED;
```

Comment:
where is messageInstance freed in this case?

Reply:
The call to `IoTHubModuleClient_LL_SendEventToOutputAsync` queues the `messageInstance` to be sent. When that happens, or fails, `SendConfirmationCallback` is called by the IoT Hub SDK. That function frees the `messageInstance`

Reply:
Suggest adding a // comment explaining where it's freed